File: nanobot-channel-webui/src/nanobot_channel_webui/case_graph/graph_state_service.py
"""Application service for canonical case graph state operations."""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any

from ..storage_paths import case_graph_contexts_root
from .graph_repository import GraphRepository
from .graph_state_types import normalize_layout
from .graph_state_types import now_iso


class GraphStateService:
    def __init__(self, workspace_root: Path) -> None:
        self._repository = GraphRepository(workspace_root)

    def load_current(self, case_id: str, graph_id: str) -> dict[str, Any]:
        return self._repository.load_current(case_id, graph_id)

    def list_steps(self, case_id: str, graph_id: str) -> list[dict[str, Any]]:
        return self._repository.list_steps(case_id, graph_id)

    def update_graph_settings(
        self,
        *,
        case_id: str,
        graph_id: str,
        settings: dict[str, Any],
    ) -> dict[str, Any]:
        return self._repository.update_graph_settings(
            case_id=case_id,
            graph_id=graph_id,
            settings=settings,
        )

    def update_layout(
        self,
        *,
        case_id: str,
        graph_id: str,
        graph_name: str = "",
        node_positions: dict[str, Any],
        position_meta: dict[str, Any] | None = None,
        group_layout: dict[str, Any] | None = None,
        viewport: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        current = self._repository.load_current(case_id, graph_id)
        graph = dict(current["graph"])
        graph_node_ids = {
            str(node.get("id") or "").strip()
            for node in graph.get("nodes") or []
            if isinstance(node, dict) and str(node.get("id") or "").strip()
        }
        group_ids = {
            str(group.get("id") or "").strip()
            for group in graph.get("investigationGroups") or []
            if isinstance(group, dict) and str(group.get("id") or "").strip()
        }
        graph_node_positions: dict[str, dict[str, float]] = {}
        group_positions: dict[str, dict[str, float]] = {}
        for raw_node_id, raw_point in node_positions.items():
            node_id = str(raw_node_id or "").strip()
            if not node_id or not isinstance(raw_point, dict):
                continue
            try:
                x = float(raw_point.get("x"))
                y = float(raw_point.get("y"))
            except (TypeError, ValueError):
                continue
            if not math.isfinite(x) or not math.isfinite(y):
                continue
            if node_id in graph_node_ids:
                graph_node_positions[node_id] = {"x": x, "y": y}
            elif node_id in group_ids:
                group_positions[node_id] = {"x": x, "y": y}

        next_groups = []
        for group in graph.get("investigationGroups") or []:
            if not isinstance(group, dict):
                continue
            group_id = str(group.get("id") or "").strip()
            position = group_positions.get(group_id)
            next_groups.append({**group, **position} if position else group)

        previous_layout = dict(graph.get("layout") or {})
        next_layout = normalize_layout({
            "nodePositions": graph_node_positions,
            "positionMeta": dict(position_meta or previous_layout.get("positionMeta") or {}),
            "groupLayout": dict(group_layout or previous_layout.get("groupLayout") or {}),
            "viewport": dict(viewport or previous_layout.get("viewport") or {}),
        }, graph.get("nodes") or [])
        if previous_layout == next_layout and next_groups == list(graph.get("investigationGroups") or []):
            return current
        graph["layout"] = next_layout
        graph["investigationGroups"] = next_groups
        state_result = self._repository.append_step(
            case_id=case_id,
            graph_id=graph_id,
            graph_name=graph_name or str(current.get("graphName") or ""),
            operation={
                "type": "update_layout",
                "label": "更新布局",
                "params": {"changedNodeCount": len(node_positions)},
            },
            graph=graph,
            delta={
                "addedNodes": [],
                "addedEdges": [],
                "updatedNodes": [{"id": node_id} for node_id in node_positions],
                "updatedEdges": [],
                "removedNodes": [],
                "removedEdges": [],
            },
            summary={"changedNodeCount": len(node_positions)},
        )
        return state_result["graph"]

    def update_node_note(
        self,
        *,
        case_id: str,
        graph_id: str,
        graph_name: str = "",
        node_id: str,
        note: str = "",
        source_note: str = "",
    ) -> dict[str, Any]:
        normalized_node_id = str(node_id or "").strip()
        if not normalized_node_id:
            raise ValueError("missing node_id")
        current = self._repository.load_current(case_id, graph_id)
        graph = dict(current["graph"])
        nodes = []
        changed = False
        for node in graph.get("nodes") or []:
            if not isinstance(node, dict):
                continue
            next_node = dict(node)
            if str(next_node.get("id") or "").strip() == normalized_node_id:
                next_node["note"] = str(note or "").strip()
                next_node["sourceNote"] = str(source_note or "").strip()
                changed = next_node != node
            nodes.append(next_node)
        if not any(str(node.get("id") or "").strip() == normalized_node_id for node in nodes):
            raise KeyError(normalized_node_id)
        if not changed:
            return current
        graph["nodes"] = nodes
        state_result = self._repository.append_step(
            case_id=case_id,
            graph_id=graph_id,
            graph_name=graph_name or str(current.get("graphName") or ""),
            operation={
                "type": "node_note_update",
                "label": "标注主体备注",
                "params": {"nodeId": normalized_node_id},
            },
            graph=graph,
            delta={
                "addedNodes": [],
                "addedEdges": [],
                "updatedNodes": [{"id": normalized_node_id}],
                "updatedEdges": [],
                "removedNodes": [],
                "removedEdges": [],
            },
            summary={"updatedNodeCount": 1},
        )
        return state_result["graph"]

    def update_latest_step_layout(
        self,
        *,
        case_id: str,
        graph_id: str,
        node_positions: dict[str, Any],
        position_meta: dict[str, Any] | None = None,
        group_layout: dict[str, Any] | None = None,
        viewport: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self._repository.update_latest_step_layout(
            case_id=case_id,
            graph_id=graph_id,
            node_positions=node_positions,
            position_meta=position_meta,
            group_layout=group_layout,
            viewport=viewport,
        )

    def write_current_context(
        self,
        *,
        case_id: str,
        graph_id: str,
        graph_name: str = "",
        chat_id: str = "",
        focus: dict[str, Any] | None = None,
        latest_step_id: str = "",
        latest_operation: dict[str, Any] | None = None,
        latest_step_summary: dict[str, Any] | None = None,
        delta_summary: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        current = self._repository.load_current(case_id, graph_id)
        graph = dict(current.get("graph") or {})
        steps = self._repository.list_steps(case_id, graph_id)
        latest_step = self._resolve_latest_step(steps, latest_step_id or str(current.get("lastStepId") or ""))
        graph_dir = self._repository.graph_dir(case_id, graph_id)
        context_path = (
            case_graph_contexts_root(self._repository._workspace_root)
            / self._safe_path_segment(case_id)
            / self._safe_path_segment(graph_id)
            / "current_context.json"
        )
        operation = latest_operation or (dict(latest_step.get("operation") or {}) if latest_step else {})
        summary = latest_step_summary or (dict(latest_step.get("summary") or {}) if latest_step else {})
        delta = dict(latest_step.get("delta") or {}) if latest_step else {}
        payload: dict[str, Any] = {
            "updatedAt": now_iso(),
            "graphId": graph_id,
            "caseId": case_id,
            "graphName": str(current.get("graphName") or graph_name or "").strip(),
            "graphFile": str((graph_dir / "graph.json").resolve()),
            "stepsDir": str((graph_dir / "steps").resolve()),
            "tradeFactsFile": str((graph_dir / "facts" / "trades.jsonl").resolve()),
            "contextFile": str(context_path.resolve()),
            "chatId": str(chat_id or "").strip(),
            "focus": dict(focus) if focus is not None else None,
            "latestStepId": str((latest_step or {}).get("stepId") or latest_step_id or current.get("lastStepId") or ""),
            "latestStepFile": str((latest_step or {}).get("file") or ""),
            "latestOperation": operation,
            "latestStepSummary": summary,
            "deltaSummary": delta_summary or self._delta_summary(delta),
            "graphStats": {
                "nodeCount": len(graph.get("nodes") or []),
                "edgeCount": len(graph.get("edges") or []),
                "tradeFactCount": len(graph.get("tradeFacts") or {}),
                "manualTradeCount": len(graph.get("manualEdges") or []),
                "realityRelationCount": len(graph.get("realityRelations") or []),
                "excludedNodeCount": len(graph.get("excludedNodes") or []),
                "excludedTradeCount": len(graph.get("excludedTrades") or []),
            },
            "availableActions": self._available_actions(graph),
        }
        self._write_json(context_path, payload)
        return payload

    @staticmethod
    def _resolve_latest_step(steps: list[dict[str, Any]], step_id: str) -> dict[str, Any] | None:
        if not steps:
            return None
        normalized_step_id = str(step_id or "").strip()
        if normalized_step_id:
            for step in steps:
                if str(step.get("stepId") or "").strip() == normalized_step_id:
                    return step
        return sorted(
            steps,
            key=lambda item: (
                int(item.get("revision") or 0),
                str(item.get("stepId") or ""),
            ),
        )[-1]

    @staticmethod
    def _delta_summary(delta: dict[str, Any]) -> dict[str, int]:
        return {
            "addedNodeCount": len(delta.get("addedNodes") or []),
            "addedEdgeCount": len(delta.get("addedEdges") or []),
            "updatedNodeCount": len(delta.get("updatedNodes") or []),
            "updatedEdgeCount": len(delta.get("updatedEdges") or []),
            "removedNodeCount": len(delta.get("removedNodes") or []),
            "removedEdgeCount": len(delta.get("removedEdges") or []),
        }

    @staticmethod
    def _available_actions(graph: dict[str, Any]) -> list[str]:
        actions = ["总结整图"]
        if graph.get("nodes"):
            actions.extend(["上钻", "下钻", "双向钻取", "全图筛选", "取消上图", "线索扩展"])
        if graph.get("edges"):
            actions.extend(["查看交易明细", "补全图上关系"])
        if graph.get("excludedNodes"):
            actions.append("恢复排除节点")
        return actions

    @staticmethod
    def _safe_path_segment(value: str) -> str:
        normalized = "".join(ch if ch.isalnum() or ch in "._-" else "-" for ch in value.strip())
        normalized = normalized.strip(".-")
        return normalized or "unknown"

    @staticmethod
    def _write_json(path: Path, payload: dict[str, Any]) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        temp_path = path.with_suffix(".json.tmp")
        temp_path.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
        temp_path.replace(path)

File: casework-node/packages/case-graph/src/graph-state-service.ts
import { mkdir, rename, writeFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { GraphRepository } from "./graph-repository.js";
import { normalizeLayout, record, text, type JsonRecord } from "./state.js";
import { safeSegment } from "./case-graph-storage.js";

export class GraphStateService {
  readonly repository: GraphRepository;
  constructor(
    caseGraphsRoot: string,
    private readonly contextRoot: string,
  ) {
    this.repository = new GraphRepository(caseGraphsRoot);
  }
  loadCurrent(caseId: string, graphId: string) {
    return this.repository.loadCurrent(caseId, graphId);
  }
  listSteps(caseId: string, graphId: string) {
    return this.repository.listSteps(caseId, graphId);
  }
  updateGraphSettings(caseId: string, graphId: string, settings: JsonRecord) {
    return this.repository.updateGraphSettings(caseId, graphId, settings);
  }
  updateLatestStepLayout(caseId: string, graphId: string, options: JsonRecord) {
    return this.repository.updateLatestStepLayout(caseId, graphId, options);
  }

  async updateLayout(input: {
    caseId: string;
    graphId: string;
    graphName?: string;
    nodePositions: JsonRecord;
    positionMeta?: JsonRecord;
    groupLayout?: JsonRecord;
    viewport?: JsonRecord;
  }): Promise<JsonRecord> {
    const current = await this.repository.loadCurrent(
      input.caseId,
      input.graphId,
    );
    const graph = record(current.graph);
    const nodeIds = new Set(
      (Array.isArray(graph.nodes) ? graph.nodes : [])
        .map((item: unknown) => text(record(item).id))
        .filter(Boolean),
    );
    const groupIds = new Set(
      (Array.isArray(graph.investigationGroups)
        ? graph.investigationGroups
        : []
      )
        .map((item: unknown) => text(record(item).id))
        .filter(Boolean),
    );
    const nodePositions: JsonRecord = {};
    const groupPositions: JsonRecord = {};
    for (const [id, rawPoint] of Object.entries(input.nodePositions)) {
      const point = record(rawPoint);
      const x = Number(point.x);
      const y = Number(point.y);
      if (!id || !Number.isFinite(x) || !Number.isFinite(y)) continue;
      if (nodeIds.has(id)) nodePositions[id] = { x, y };
      else if (groupIds.has(id)) groupPositions[id] = { x, y };
    }
    const previousGroups = Array.isArray(graph.investigationGroups)
      ? graph.investigationGroups
      : [];
    const nextGroups = previousGroups.map((raw: unknown) => {
      const group = record(raw);
      return groupPositions[text(group.id)]
        ? { ...group, ...record(groupPositions[text(group.id)]) }
        : group;
    });
    const previousLayout = record(graph.layout);
    const nextLayout = normalizeLayout(
      {
        nodePositions,
        positionMeta: input.positionMeta ?? previousLayout.positionMeta ?? {},
        groupLayout: input.groupLayout ?? previousLayout.groupLayout ?? {},
        viewport: input.viewport ?? previousLayout.viewport ?? {},
      },
      graph.nodes as JsonRecord[],
    );
    if (
      JSON.stringify(previousLayout) === JSON.stringify(nextLayout) &&
      JSON.stringify(previousGroups) === JSON.stringify(nextGroups)
    )
      return current;
    graph.layout = nextLayout;
    graph.investigationGroups = nextGroups;
    const result = await this.repository.appendStep({
      caseId: input.caseId,
      graphId: input.graphId,
      graphName: input.graphName || text(current.graphName),
      operation: {
        type: "update_layout",
        label: "更新布局",
        params: { changedNodeCount: Object.keys(input.nodePositions).length },
      },
      graph,
      delta: {
        addedNodes: [],
        addedEdges: [],
        updatedNodes: Object.keys(input.nodePositions).map((id) => ({ id })),
        updatedEdges: [],
        removedNodes: [],
        removedEdges: [],
      },
      summary: { changedNodeCount: Object.keys(input.nodePositions).length },
    });
    return record(result.graph);
  }

  async updateNodeNote(input: {
    caseId: string;
    graphId: string;
    graphName?: string;
    nodeId: string;
    note?: string;
    sourceNote?: string;
  }): Promise<JsonRecord> {
    const nodeId = text(input.nodeId);
    if (!nodeId) throw new Error("missing node_id");
    const current = await this.repository.loadCurrent(
      input.caseId,
      input.graphId,
    );
    const graph = record(current.graph);
    let found = false;
    let changed = false;
    graph.nodes = (Array.isArray(graph.nodes) ? graph.nodes : []).map(
      (raw: unknown) => {
        const node = record(raw);
        if (text(node.id) !== nodeId) return node;
        found = true;
        const next = {
          ...node,
          note: text(input.note),
          sourceNote: text(input.sourceNote),
        };
        changed = JSON.stringify(next) !== JSON.stringify(node);
        return next;
      },
    );
    if (!found) throw new Error(`node '${nodeId}' not found`);
    if (!changed) return current;
    const result = await this.repository.appendStep({
      caseId: input.caseId,
      graphId: input.graphId,
      graphName: input.graphName || text(current.graphName),
      operation: {
        type: "node_note_update",
        label: "标注主体备注",
        params: { nodeId },
      },
      graph,
      delta: {
        addedNodes: [],
        addedEdges: [],
        updatedNodes: [{ id: nodeId }],
        updatedEdges: [],
        removedNodes: [],
        removedEdges: [],
      },
      summary: { updatedNodeCount: 1 },
    });
    return record(result.graph);
  }

  async writeCurrentContext(input: {
    caseId: string;
    graphId: string;
    graphName?: string;
    chatId?: string;
    focus?: JsonRecord | null;
    latestStepId?: string;
    latestOperation?: JsonRecord;
    latestStepSummary?: JsonRecord;
    deltaSummary?: JsonRecord;
  }): Promise<JsonRecord> {
    const current = await this.repository.loadCurrent(
      input.caseId,
      input.graphId,
    );
    const graph = record(current.graph);
    const steps = await this.repository.listSteps(input.caseId, input.graphId);
    const wanted = text(input.latestStepId || current.lastStepId);
    const latest =
      steps.find((step) => text(step.stepId) === wanted) ??
      [...steps]
        .sort(
          (a, b) =>
            Number(a.revision || 0) - Number(b.revision || 0) ||
            text(a.stepId).localeCompare(text(b.stepId)),
        )
        .at(-1);
    const graphDir = this.repository.graphDir(input.caseId, input.graphId);
    const contextFile = join(
      this.contextRoot,
      safeSegment(input.caseId),
      safeSegment(input.graphId),
      "current_context.json",
    );
    const delta = record(latest?.delta);
    const payload = {
      updatedAt: new Date().toISOString(),
      graphId: input.graphId,
      caseId: input.caseId,
      graphName: text(current.graphName || input.graphName),
      graphFile: resolve(join(graphDir, "graph.json")),
      stepsDir: resolve(join(graphDir, "steps")),
      tradeFactsFile: resolve(join(graphDir, "facts", "trades.jsonl")),
      contextFile: resolve(contextFile),
      chatId: text(input.chatId),
      focus: input.focus ? record(input.focus) : null,
      latestStepId: text(
        latest?.stepId || input.latestStepId || current.lastStepId,
      ),
      latestStepFile: text(latest?.file),
      latestOperation: input.latestOperation ?? record(latest?.operation),
      latestStepSummary: input.latestStepSummary ?? record(latest?.summary),
      deltaSummary: input.deltaSummary ?? deltaSummary(delta),
      graphStats: {
        nodeCount: arrayLength(graph.nodes),
        edgeCount: arrayLength(graph.edges),
        tradeFactCount: Object.keys(record(graph.tradeFacts)).length,
        manualTradeCount: arrayLength(graph.manualEdges),
        realityRelationCount: arrayLength(graph.realityRelations),
        excludedNodeCount: arrayLength(graph.excludedNodes),
        excludedTradeCount: arrayLength(graph.excludedTrades),
      },
      availableActions: availableActions(graph),
    };
    await atomicJson(contextFile, payload);
    return payload;
  }
}

function arrayLength(value: unknown): number {
  return Array.isArray(value) ? value.length : 0;
}
function deltaSummary(delta: JsonRecord) {
  return {
    addedNodeCount: arrayLength(delta.addedNodes),
    addedEdgeCount: arrayLength(delta.addedEdges),
    updatedNodeCount: arrayLength(delta.updatedNodes),
    updatedEdgeCount: arrayLength(delta.updatedEdges),
    removedNodeCount: arrayLength(delta.removedNodes),
    removedEdgeCount: arrayLength(delta.removedEdges),
  };
}
function availableActions(graph: JsonRecord): string[] {
  const result = ["总结整图"];
  if (arrayLength(graph.nodes))
    result.push("上钻", "下钻", "双向钻取", "全图筛选", "取消上图", "线索扩展");
  if (arrayLength(graph.edges)) result.push("查看交易明细", "补全图上关系");
  if (arrayLength(graph.excludedNodes)) result.push("恢复排除节点");
  return result;
}
async function atomicJson(path: string, value: unknown): Promise<void> {
  await mkdir(join(path, ".."), { recursive: true });
  const temp = `${path}.tmp`;
  await writeFile(temp, JSON.stringify(value, null, 2), "utf8");
  await rename(temp, path);
}
